### PR TITLE
Bulk Update for Postgres

### PR DESCRIPTION
**Pull request by suddendust**
https://github.com/hypertrace/document-store/issues/82

This change adds support for `bulkUpdate` in Postgres.

### CODE REVIEW

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/postgres/PostgresDocStoreTest.java [hunk 1]
```diff
@@ -162,4 +176,192 @@ public void testDrop() {
     collection.drop();
     Assertions.assertFalse(datastore.listCollections().contains("postgres." + COLLECTION_NAME));
   }
+
+  @Test
+  public void whenBulkUpdatingExistingRecords_thenExpectOnlyRecordsWhoseConditionsMatchToBeUpdated()
```
Comment (kotharironak) [inline document-store/src/integrationTest/java/org/hypertrace/core/documentstore/postgres/PostgresDocStoreTest.java:181]: Pl. move this to DocStoreTest to run against both mongo + postgres.
Prior, it was part of MongoDocStoreTest as postgres was not supporting impl for this API. Remove it from `MongoDocStoreTest`

Comment (kotharironak) [reply document-store/src/integrationTest/java/org/hypertrace/core/documentstore/postgres/PostgresDocStoreTest.java:181]: Move all the `collection.bulkUpdate` test from `MongoDocStoreTest` and move to `DocStoreTest`

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/postgres/PostgresDocStoreTest.java [hunk 2]
```diff
@@ -1,5 +1,6 @@
 package org.hypertrace.core.documentstore.postgres;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
```
Comment (kotharironak) [inline document-store/src/integrationTest/java/org/hypertrace/core/documentstore/postgres/PostgresDocStoreTest.java:3]: Looks like no change

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java [hunk 3]
```diff
@@ -130,7 +130,26 @@ public CreateResult create(Key key, Document document) throws IOException {
 
   @Override
   public BulkUpdateResult bulkUpdate(List<BulkUpdateRequest> bulkUpdateRequests) throws Exception {
-    throw new UnsupportedOperationException();
+
+    var requestsWithFilter =
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java:134]: nit: replace `var` with actual types for readability.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java [hunk 4]
```diff
@@ -130,7 +130,26 @@ public CreateResult create(Key key, Document document) throws IOException {
...
+        bulkUpdateRequests.stream()
+            .filter(req -> req.getFilter() != null)
+            .collect(Collectors.toList());
+
+    var requestsWithoutFilter =
+        bulkUpdateRequests.stream()
+            .filter(req -> req.getFilter() == null)
+            .collect(Collectors.toList());
+
+    int totalUpdateCountA = bulkUpdateRequestsWithFilter(requestsWithFilter);
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java:144]: As discussed, let's go with the late exception approach.
- e.g if there are 10 documents to update, we will try for each document even if there is an expectation say updating `3rd` document.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java [hunk 5]
```diff
@@ -327,6 +346,25 @@ protected PreparedStatement buildPreparedStatement(String sqlQuery, Params param
     return preparedStatement;
   }
 
+  @VisibleForTesting
+  protected void enrichPreparedStatementWithParams(
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java:350]: Use this new method inside `buildPreparedStatement`

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java (hunk 4 above)
Comment (suddendust) [reply document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java]: Done

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java (hunk 3 above)
Comment (suddendust) [reply document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java]: Done

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java [hunk 6]
```diff
@@ -546,6 +581,89 @@ private int[] bulkUpsertImpl(Map<Key, Document> documents) throws SQLException,
...
+    long totalRowsUpdated = 0;
+    Exception sampleException = null;
+
+    for (BulkUpdateRequest request : requests) {
+      Key key = request.getKey();
+      Document document = request.getDocument();
+      Filter filter = request.getFilter();
+
+      try {
+        totalRowsUpdated += update(key, document, filter).getUpdatedCount();
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java:597]: As a follow-up, we can explore `java.sql.Statement.addBatch` - https://github.com/JetBrains/jdk8u_jdk/blob/master/src/share/classes/java/sql/Statement.java#L520

However, its not on PreparedStatement.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java [hunk 7]
```diff
@@ -134,7 +134,36 @@ public CreateResult create(Key key, Document document) throws IOException {
...
+            .filter(req -> req.getFilter() != null)
+            .collect(Collectors.toList());
+
+    List<BulkUpdateRequest> requestsWithoutFilter =
+        bulkUpdateRequests.stream()
+            .filter(req -> req.getFilter() == null)
+            .collect(Collectors.toList());
+
+    try {
+      long totalUpdateCountA = bulkUpdateRequestsWithFilter(requestsWithFilter);
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java:149]: I thought we were going with the late exception as discussed yesterday. We can take care of it as follow-up PR.

**Review by kotharironak — APPROVED** [review]
Overall look fine, we can take care of comments in follow-up. Merging this one.

**Pull request merged**